### PR TITLE
Server spinoffs

### PR DESCRIPTION
**Pull request by clm805**

### CODE REVIEW

File: lib/pirates/game_server/manager.ex [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  # handle dropped servers
+  def handle_info({:DOWN, _ref, :process, pid, _reason}, servers) do
+      servers = List.delete(servers, pid)
+    {:noreply, servers}
+  end
+
+  defp count(s) do
+      Instance.count(Instance.table(s))
```
Comment (SylvanSign) [inline lib/pirates/game_server/manager.ex]: can you use `server` rather than `s`. It's unclear what `s` means

File: web/channels/game_channel.ex [hunk 2]
```diff
@@ -1,12 +1,15 @@
 defmodule Pirates.GameChannel do
   use Phoenix.Channel
-  alias Pirates.GameServer.State
+  alias Pirates.GameServer.Instance.State
+  require Logger
```
Comment (SylvanSign) [inline web/channels/game_channel.ex]: not logging, so remove this

File: web/channels/game_channel.ex [hunk 3]
```diff
@@ -1,12 +1,15 @@
...
-    {:ok, table} = Pirates.GameServer.register()
-    socket = assign(socket, :table, table)
-    socket = assign(socket, :id, System.unique_integer([:positive]))
-    {:ok, socket}
+    {:ok, server} = Pirates.GameServer.Manager.get_available_server()
+    {:ok, table} = Pirates.GameServer.Instance.register(server)
+    
+    {:ok, socket |>
+      assign(:table, table) |> 
+      assign(:id, System.unique_integer([:positive]))}
```
Comment (SylvanSign) [inline web/channels/game_channel.ex]: pls fix pipe styling as before

also, probably cleaner to do
```elixir
socket = socket
  |> foo()
  ...
{:ok, socket}

File: lib/pirates/game_server/instance.ex [hunk 4]
```diff
@@ -115,7 +115,7 @@ defmodule Pirates.GameServer.Instance do
     :ets.delete(table, pid)
     IO.puts "Just deleted state for #{inspect(pid)}"
     if count(table) == 0 do
-      Process.exit(self(), "Server was empty.")
+      Process.exit("Server was empty.")
```
Comment (SylvanSign) [inline lib/pirates/game_server/instance.ex]: it's just `exit/1`, which is part of the automatically imported `Kernel` module

File: web/channels/game_channel.ex (hunk 3 above)
Comment (clm805) [inline web/channels/game_channel.ex]: I'd rather not assign a var just to return it. Would rather just return it.

File: lib/pirates/game_server/manager.ex [hunk 5]
```diff
@@ -17,7 +17,7 @@ defmodule Pirates.GameServer.Manager do
         GenServer.start_link(@name, :ok, name: @name)
     end
 
-    def get_available_server do
+    def next_available_server do
         GenServer.call(@name, :next)
```
Comment (SylvanSign) [inline lib/pirates/game_server/manager.ex]: cab you make the message tag the same as the function call pls

File: lib/pirates/game_server/manager.ex [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  ##############
+  # Client API #
+  ##############
+
+    def start_link do
+        GenServer.start_link(@name, :ok, name: @name)
+    end
+
+    def next_available_server do
+        GenServer.call(@name, :next)
```
Comment (SylvanSign) [inline lib/pirates/game_server/manager.ex]: can you just make the message tag `:next_available_server` as well?

File: lib/pirates/game_server/instance.ex [hunk 7]
```diff
@@ -98,10 +104,19 @@ defmodule Pirates.GameServer do
...
+    Enum.each(pids, fn pid -> send(pid, {:state_tick, states}) end)
+    {:noreply, table}
+  end
+
   # handle notifications of dead monitored processes
   def handle_info({:DOWN, _ref, :process, pid, _reason}, table) do
     :ets.delete(table, pid)
     IO.puts "Just deleted state for #{inspect(pid)}"
+    if count(table) == 0 do
+      exit(:shutdown)
```
Comment (SylvanSign) [inline lib/pirates/game_server/instance.ex]: I think the _actual_ best practice is to return `{:stop, :shutdown, table}` here

> Returning `{:stop, reason, new_state}` stops the loop and terminate/2 is called with the reason reason and state new_state. The process exits with reason reason.

So maybe replace this and the next line with
```elixir
case count(table) do
  0 -> {:stop, :shutdown, table}
  _ -> {:noreply, table}
end

File: lib/pirates/game_server/manager.ex [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    server = 
+        servers 
+        |> Enum.filter(fn (s) -> count(s) < @max_players_per_server end) 
+        |> Enum.min_by(&count/1, fn -> create_server() end)
+    {:reply, {:ok, server}, MapSet.put(servers, server)}
+  end
+
+  # handle dropped servers
+  def handle_info({:DOWN, _ref, :process, pid, _reason}, servers) do
+    {:noreply, MapSet.delete(servers, pid)}
```
Comment (SylvanSign) [inline lib/pirates/game_server/manager.ex]: can we rename `pid` to `server` here to be consistent with the rest of the module?

File: lib/pirates/game_server/manager.ex [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  defp count(s) do
+      Instance.count(Instance.table(s))
```
Comment (SylvanSign) [inline lib/pirates/game_server/manager.ex]:
```elixir
s |> Instance.table() |> Instance.count()

File: web/channels/game_channel.ex (hunk 3 above)
Comment (SylvanSign) [inline web/channels/game_channel.ex]: Your complaint isn't in line with several other bits of code you've already written, plus it's easier to read with the temp variable.

File: lib/pirates/game_server/manager.ex [hunk 10]
```diff
@@ -39,12 +39,12 @@ defmodule Pirates.GameServer.Manager do
...
-  def handle_info({:DOWN, _ref, :process, pid, _reason}, servers) do
-    {:noreply, MapSet.delete(servers, pid)}
+  def handle_info({:DOWN, _ref, :process, server, _reason}, servers) do
+    {:noreply, MapSet.delete(servers, server)}
   end
 
-  defp count(s) do
-      Instance.count(Instance.table(s))
+  defp count(server) do
+      server |> Instance.table |> Instance.count
```
Comment (SylvanSign) [inline lib/pirates/game_server/manager.ex]: use Parens pls https://github.com/christopheradams/elixir_style_guide#parentheses-pipe-operator

File: lib/pirates/game_server/manager.ex [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    {:reply, {:ok, server}, MapSet.put(servers, server)}
+  end
+
+  # handle dropped servers
+  def handle_info({:DOWN, _ref, :process, server, _reason}, servers) do
+    {:noreply, MapSet.delete(servers, server)}
+  end
+
+  defp count(server) do
+      server |> Instance.table |> Instance.count
```
Comment (SylvanSign) [inline lib/pirates/game_server/manager.ex]: can you use parens always in pipes?

Comment (SylvanSign) [inline lib/pirates/game_server/manager.ex]: https://github.com/christopheradams/elixir_style_guide#parentheses-pipe-operator

**Pull request merged**